src/client_features.rs

```diff
@@ -52,117 +52,97 @@ pub enum Operator {
-// The second reason is that we can't shield the Rust code from consumers using the FFI layers and potentially doing
-// exactly the same thing in languages that allow it. They should not do that. But if they do we have enough information
-// to understand the intent of the executed code clearly and there's no reason to fail.
-// This also maps numbers + booleans to strings, and disregards other types without failing
-fn stringify_numbers_and_booleans_remove_nulls_and_non_strings<'de, D>(
-    deserializer: D,
-) -> Result<Option<HashMap<String, String>>, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    let props: Option<HashMap<String, Option<Value>>> = Option::deserialize(deserializer)?;
-    Ok(props.map(|props| {
-        props
-            .into_iter()
-            .filter_map(|(k, v)| match v {
-                Some(Value::String(s)) => {
-                    if s.is_empty() {
-                        None
-                    } else {
-                        Some((k, s))
-                    }
-                }
-                Some(Value::Number(n)) => Some((k, n.to_string())),
-                Some(Value::Bool(b)) => Some((k, b.to_string())),
-                _ => None,
-            })
-            .collect()
-    }))
-}
+impl<'de> Deserialize<'de> for Context {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let mut raw: Map<String, Value> = Deserialize::deserialize(deserializer)?;
 
-// Allowing a looser deserialization for the contexts properties to match Unleash behavior
-fn stringify_numbers_and_booleans<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    let prop: Option<Value> = Option::deserialize(deserializer)?;
-    Ok(match prop {
-        Some(Value::String(s)) => {
-            if s.is_empty() {
-                None
+        if let Some(context_val) = raw.remove("context") {
+            if let Value::Object(inner) = context_val {
+                return Context::from_map(inner);
             } else {
-                Some(s)
+                return Err(serde::de::Error::custom(
+                    "Expected 'context' to be an object",
+                ));
             }
         }
-        Some(Value::Number(n)) => Some(n.to_string()),
-        Some(Value::Bool(b)) => Some(b.to_string()),
-        _ => None,
-    })
+
+        Context::from_map(raw)
+    }
+}
+
+impl Context {
+    fn from_map<E: serde::de::Error>(mut raw: Map<String, Value>) -> Result<Self, E> {
+        fn parse_value(v: Value) -> Option<String> {
+            match v {
+                Value::String(s) => Some(s),
+                Value::Number(n) => Some(n.to_string()),
+                Value::Bool(b) => Some(b.to_string()),
+                _ => None,
+            }
+        }
+
+        fn extract_property(
```

Comment:
We use this to extract a given property from the context. So if `userId` is present either at a base level or on the properties, we move it to the base level

You guessed it - that's what the frontend API does